x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/CountingRunner.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.security.support;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.common.cache.Cache;
+import org.elasticsearch.common.util.concurrent.ReleasableLock;
+
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+/**
+ * An utility class that keeps a counter when executing runnables.
+ * It is designed to help minimizing the possibility of caching stable results in a {@link Cache}.
+ */
+public class CountingRunner {
+
+    private static final Logger logger = LogManager.getLogger(CountingRunner.class);
+
+    private final AtomicLong counter = new AtomicLong();
+    private final ReadWriteLock countingLock = new ReentrantReadWriteLock();
+    private final ReleasableLock countingReadLock = new ReleasableLock(countingLock.readLock());
+    private final ReleasableLock countingWriteLock = new ReleasableLock(countingLock.writeLock());
+
+    public long getCount() {
+        return counter.get();
+    }
+
+    /**
+     * Execute the given runnable if the internal counter matches the given count.
+     * The counter check is performed inside a read-locking block to prevent incrementing of
+     * the counter, i.e. call to {@link CountingRunner#incrementAndRun} will be blocked.
+     * But it does *not* block other invocations of {@link CountingRunner#runIfCountMatches}.
+     *
+     * @return true if the runnable is executed, other false.
+     */
+    public boolean runIfCountMatches(Runnable runnable, long count) {
```

Comment:
I think it would be easier to read if the `count` were the first parameter.

Reply:
Maybe rename to `compareAndRun`? I think this name helps with the intuition from "compare and swap".
"counter" is used too much so maybe drop it if there's a parameter with the name counter already.

Reply:
`Runnable` -> `CheckedRunnable`, to avoid worrying about the checked/unchecked exceptions at this low level.

Reply:
Your analogy of `AtomicInteger` makes sense. The name `compareAndRun` also sounds good and to the point. The `count` parameter naturally should become the first parameter after the method name change to follow the convention of `compareAndSet`. Thanks!

I decided to keep `Runnable` instead of changing it to `CheckedRunnable`. In fact, I started with `CheckedRunnable`, but I think it has a few disadvantages:
1. If we don't handle the exception of `CheckedRunnable`, the method would have a signature like:
  `public <E extends Exception> boolean compareAndRun(long count, CheckedRunnable<E> runnable) throws E`
  It forces its caller to handle exceptions even when the Runnable may not throw any exception (as in the `NativePrivilegesStore` case)
2. If we handle the exception inside the method and only bubble up `RuntimeException`, it will make the method easier to call because no exception handling is forced onto the caller. However, the method itself has no knowledge about he Runnable so it cannot perform meaningful handling other than maybe wrap it with a RuntimeException. It works, but makes it hard if the caller cares about the real exception and wants to handle it.

My understanding is that `CheckedRunnable` is better when possible exception types are limited and the method knows well how to handle (or ignore) them. Therefore, I prefer to just using `Runnable`. It works similar to above item 2, but giving caller full control of how exception should or should not be handled.

Reply:
Okay, sounds good to me. It's not going to be a big difference in practice anyway.